File: AppServiceBlinky/C#/BlinkyClient/StartupTask.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Http;
using Windows.ApplicationModel.Background;
using Windows.ApplicationModel.AppService;
using Windows.Foundation.Collections;
using Windows.System.Threading;


// The Background Application template is documented at http://go.microsoft.com/fwlink/?LinkID=533884&clcid=0x409

namespace BlinkyClient
{
    public sealed class StartupTask : IBackgroundTask
    {

        AppServiceConnection connection;
        BackgroundTaskDeferral deferral;
        ThreadPoolTimer timer;
        string requestedPinValue;


        public async void Run(IBackgroundTaskInstance taskInstance)
        {
            deferral = taskInstance.GetDeferral();


            //Connect to the "BlinkyService" implemented in the "BlinkyService" solution
            connection = new AppServiceConnection();
            connection.AppServiceName = "BlinkyService";
            connection.PackageFamilyName = "BlinkyService-uwp_gpek5j0d8wyr0";
            AppServiceConnectionStatus status = await connection.OpenAsync();

            if (status != AppServiceConnectionStatus.Success)
            {
                deferral.Complete();
                return;
            }

            //Send a message with the name "requestedPinValue" and the value "High"
            //These work like loosely typed input parameters to a method
            requestedPinValue = "High";
            var message = new ValueSet();
            message["requestedPinValue"] = requestedPinValue;
            AppServiceResponse response = await connection.SendMessageAsync(message);

            //If the message was successful, start a timer to send alternating requestedPinValues to blink the LED
            if (response.Status == AppServiceResponseStatus.Success)
            {
                timer = ThreadPoolTimer.CreatePeriodicTimer(this.Tick, TimeSpan.FromMilliseconds(500));
            }

        }

        private async void Tick(ThreadPoolTimer timer)
        {
            if (requestedPinValue.Equals("High"))
            {
                requestedPinValue = "Low";
            }
            else
            {
                requestedPinValue = "High";
            }
            var message = new ValueSet();
            message["requestedPinValue"] = requestedPinValue;
            await connection.SendMessageAsync(message);
        }
    }
}

File: AppServiceBlinky/Cpp/BlinkyClient/StartupTask.cpp
#include "pch.h"
#include "StartupTask.h"

using namespace BlinkyClient;

using namespace Platform;
using namespace Windows::ApplicationModel::Background;
using namespace Windows::ApplicationModel::AppService;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::System::Threading;
using namespace Windows::Devices::Gpio;

// The Background Application template is documented at http://go.microsoft.com/fwlink/?LinkID=533884&clcid=0x409

void StartupTask::Run(IBackgroundTaskInstance^ taskInstance)
{
    serviceDeferral = taskInstance->GetDeferral();

    serviceConnection = ref new AppServiceConnection();
    serviceConnection->PackageFamilyName = L"BlinkyService-uwp_2yx4q2bk84nj4";
    serviceConnection->AppServiceName = L"BlinkyService";

    auto connectTask = Concurrency::create_task(serviceConnection->OpenAsync());
    connectTask.then([this](AppServiceConnectionStatus connectStatus)
    {
        if (AppServiceConnectionStatus::Success == connectStatus)
        {
            auto message = ref new ValueSet();
            command = L"SetLedStateOn";
            message->Insert(L"SetLedState", command);
            message->Insert(L"PinNumber", 5);
            auto messageTask = Concurrency::create_task(serviceConnection->SendMessageAsync(message));

            messageTask.then([this](AppServiceResponse ^appServiceResponse)
            {
                if ((appServiceResponse->Status == AppServiceResponseStatus::Success) &&
                    (appServiceResponse->Message != nullptr) &&
                    (appServiceResponse->Message->HasKey(L"Response")))
                {
                    auto response = appServiceResponse->Message->Lookup(L"Response");
                    String^ responseMessage = safe_cast<String^>(response);
                    if (String::operator==(responseMessage, L"Success"))
                    {
                        TimerElapsedHandler ^handler = ref new TimerElapsedHandler(
                            [this](ThreadPoolTimer ^timer)
                        {
                            RequestBlinkService();
                        });

                        TimeSpan interval;
                        interval.Duration = 500 * 1000 * 10;
                        timer = ThreadPoolTimer::CreatePeriodicTimer(handler, interval);
                    }
                }
            });
        }
        else
        {
            serviceDeferral->Complete();
        }
    });
}

void StartupTask::RequestBlinkService()
{
    command = String::operator==(command, L"SetLedStateOn") ? L"SetLedStateOff" : L"SetLedStateOn";
    auto message = ref new ValueSet();
    message->Insert(L"SetLedState", command);
    message->Insert(L"PinNumber", 5);
    auto messageTask = Concurrency::create_task(serviceConnection->SendMessageAsync(message));
}

void BlinkyClient::StartupTask::OnCanceled(Windows::ApplicationModel::Background::IBackgroundTaskInstance ^sender, Windows::ApplicationModel::Background::BackgroundTaskCancellationReason reason)
{
    if (serviceDeferral != nullptr)
    {
        serviceDeferral->Complete();
    }

    if (serviceConnection != nullptr)
    {
        serviceConnection = nullptr;
    }
}